framework/db/Command.php

```diff
@@ -137,19 +137,43 @@ public function getSql()
     }
 
     /**
-     * Specifies the SQL statement to be executed.
-     * The previous SQL execution (if any) will be cancelled, and [[params]] will be cleared as well.
+     * Specifies the SQL statement to be executed. The SQL statement will be quoted using [[Connection::quoteSql()]].
+     * The previous SQL execution (if any) will be cancelled, and [[params]] will be cleared as well. See [[reset()]]
+     * for details.
+     *
      * @param string $sql the SQL statement to be set.
      * @return $this this command instance
+     * @see reset()
+     * @see cancel()
      */
     public function setSql($sql)
     {
         if ($sql !== $this->_sql) {
             $this->cancel();
+            $this->reset();
             $this->_sql = $this->db->quoteSql($sql);
-            $this->_pendingParams = [];
-            $this->params = [];
-            $this->_refreshTableName = null;
+        }
```

Comment:
`$this->_sql` is the quoted version of `$sql`. So the condition `if ($sql !== $this->_sql)` doesn't test for changes.

Is this test really required?

Reply:
it probably isn't as this check has been introduced before `quoteSql` has been created. This does not belong to this PR however. Changing it could probably break code in very rare cases.